File: HIndexII.py
#Time Complexity: O(log(n))
#Space Complexity: O(1)

class Solution:
    def hIndex(self, citations: List[int]) -> int:
        n = len(citations)
        low = 0
        high = n-1

        while low<=high:
            mid = (low+high)//2
            diff = n-mid
            if citations[mid]>=diff :
                if mid==0 or citations[mid-1]<diff+1:
                    return n-mid
                high = mid-1
            elif citations[mid]<diff:
                low= mid+1
            else:
                high = mid-1
        return n-low

File: H_IndexII.java
//Time Complexity: O(n)
//Space Complexity: O(1)

class Solution {
    public int hIndex(int[] citations) {
        int n = citations.length;
        for(int i=0; i < n; i++){
            int diff = n - i;
            if(diff <= citations[i]){
                return diff;
            }
        }
        return 0;
    }
}

//Binary Search Solution
//Time Complexity: O(log n)
//Space Complexity: O(1)

class Solution {
    public int hIndex(int[] citations) {
        int n = citations.length;
        int low =0;
        int high = n-1;
        while(low <= high){
            int mid = low + (high - low)/2;
            int diff = n - mid;
            if(diff == citations[mid]){
                return diff;
            }
            else if(diff > citations[mid]){
                low = mid +1;
            }else{
                high = mid -1;
            }
        }
        return n - low;
    }
}